src/crimson/osd/osd_operations/client_request_common.cc

```diff
@@ -11,8 +11,60 @@ namespace {
   }
 }
 
+SET_SUBSYS(osd);
+
 namespace crimson::osd {
 
+InterruptibleOperation::template interruptible_future<>
+CommonClientRequest::recover_missings(
+  Ref<PG> &pg,
+  const hobject_t& soid,
+  std::vector<snapid_t> &&snaps)
+{
+  using interruptor = InterruptibleOperation::interruptor;
+  LOG_PREFIX(CommonClientRequest::recover_missings);
+  auto fut = interruptor::now();
+  if (!pg->is_primary()) {
+    DEBUGI("process_op: Skipping do_recover_missing on non primary pg");
+    return fut;
+  }
+  if (!soid.is_head()) {
+    fut = do_recover_missing(pg, soid.get_head());
+  }
+  return seastar::do_with(
+    std::move(snaps),
+    [pg, soid, fut=std::move(fut)](auto &snaps) mutable {
+    return fut.then_interruptible([&snaps, pg, soid]() mutable {
+      return pg->obc_loader.with_obc<RWState::RWREAD>(
+        soid.get_head(),
+        [&snaps, pg, soid](auto head, auto) mutable {
+        auto oid = resolve_oid(head->get_head_ss(), soid);
+        if (!oid) {
+          ceph_assert(head->get_head_ss().clones.empty());
```

Comment:
This assert is wrong -- if the operation is a read on a snapshot, there's no reason to assume that the snapshot maps to a point where the object existed.

Reply:
Yeah, corrected.